### PR TITLE
GGML MPT-7B Implementation for CPU Inference

### PR DESCRIPTION
**Pull request by FFFiend**
**What kind of change does this PR introduce?**
Addresses #141 
Adds a Replay Strategy that uses the GGML MPT-7B (with 4-bit quantization) model to generate Action completions through inference on the CPU.

**Summary**
Previous MPT-7B implementation required 12+ GB of VRAM and as such was impossible to load on MacOS. This PR resolves that issue and model download + inference can be ran on CPU, provided sufficient memory is available.

**Checklist**
* [x] My code follows the style guidelines of OpenAdapt
* [x] I have perfomed a self-review of my code
* [x] I have made corresponding changes to the documentation (e.g. README.md, requirements.txt)

**How can your code be run and tested?**
Well, mixins aren't meant to be run directly, however instructions and further information can be found at #291

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -37,3 +37,4 @@ tqdm==4.64.0
 nicegui==1.2.16
 transformers==4.29.2
 python-dotenv==1.0.0
+ctransformers
```
Comment (KrishPatel13) [inline requirements.txt:40]: It will be great if we put this at its alphabetical place when it comes to ordering in requirements :)

File: openadapt/strategies/mixins/mpt7b.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+Usage:
+
+    class MyReplayStrategy(MPT7BReplayStrategy):
+        ...
+"""
+
+
+from openadapt.strategies.base import BaseReplayStrategy
+from openadapt.models import Recording
```
Comment (KrishPatel13) [inline openadapt/strategies/mixins/mpt7b.py:14]:
```
from ctransformers import AutoModelForCausalLM
from openadapt.models import Recording
from openadapt.strategies.base import BaseReplayStrategy
```

Reference: https://github.com/MLDSAI/OpenAdapt/pull/306#discussion_r1239163590

**Comment by FFFiend** [comment]
I see that the ```active_screenshot``` param in the ```get_next_action_event``` method is unused. Did it just end up not being necessary ?

File: openadapt/strategies/mixins/mpt7b.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        self.model_path = 'TheBloke/MPT-7B-GGML'
+        self.model_file = 'mpt-7b.ggmlv3.q4_0.bin'
+
+    def get_completion(
+        self,
+        completion: str
+    ) -> str:
+        llm = AutoModelForCausalLM.from_pretrained(
+        self.model_path, model_type='mpt',
+        model_file=self.model_file)
```
Comment (abrichr) [inline openadapt/strategies/mixins/mpt7b.py:36]: Please format with black. Also please define a constant at the top of the file:

```
MODEL_TYPE = "mpt"
```

Please use double-quotes 🙏

File: openadapt/strategies/mixins/mpt7b.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    MPT-7B Replay Strategy for inference on CPU.
+    """
+
+    def __init__(
+        self,
+        recording: Recording
+    ):
+        super.__init__(recording)
+        self.model_path = 'TheBloke/MPT-7B-GGML'
+        self.model_file = 'mpt-7b.ggmlv3.q4_0.bin'
```
Comment (abrichr) [inline openadapt/strategies/mixins/mpt7b.py:28]: Please define constants at the top of the file, and pass in as default arguments here, e.g.:

```
    def __init__(
        self,
        recording: Recording,
        model_path: str = MODEL_PATH,
        model_file: str = MODEL_FILE,
    ):
    ...
    self.model_path = model_path
    self.model_file = model_file
```

**Review by abrichr — CHANGES_REQUESTED** [review]
Thanks @FFFiend ! Can you please add a smoke test at the very minimum?

**Comment by abrichr** [comment]
Closing as part of the architecture transition. The legacy codebase has been moved to `legacy/` and frozen at v0.46.0.

If this change is still needed, please:
1. Open a PR against the relevant modular package repository
2. Or re-open this PR targeting the `legacy/` directory if it's a critical legacy fix

See PR #960 for details on the new meta-package architecture.
